### DISCUSSION TITLE
Loading all deferred columns at once when the first one is accessed

### DISCUSSION THREAD

**Discussion by antoineleclair** [sqlalchemy/sqlalchemy#10718] [Usage Questions] [answered]
We're on SQLAlchemy 1.4.49 (but are planning to update to version 2), Python 3.11.6.

I'll explain my use case with simplified code example to show what I'm trying to achieve.

The models look like this:
```python
class ModelOne(Base):
    __tablename__ = "model_one"
    id = Column(String(32), primary_key=True)
    model_two_id = Column(
        String(32),
        ForeignKey("model_two.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    model_two = relationship("ModelTwo")

class ModelTwo(Base):
    __tablename__ = "model_two"
    id = Column(String(32), primary_key=True)
    column_one = Column(String(32))
    column_two = Column(String(32))
    column_three = Column(String(32))
    column_four = Column(String(32))
```

With a query that looks like this:
```python
model_one = (
    session.query(ModelOne)
    .filter(ModelOne.id == model_one_id)
    .options(
        joinedload(ModelOne.model_two).options(
            load_only(
                ModelTwo.id,
                ModelTwo.column_one,
                ModelTwo.column_two,
            )
        )
    )
    .first()
)
```

The following line will not run any SQL query because the data is already loaded:
```python
print(model_one.model_two.column_one)
```

If we run this
```python
print(model_one.model_two.column_three)
```

it will run the SQL to load that deferred column
```sql
SELECT model_two.column_three AS model_two_column_three 
FROM model_two 
WHERE model_two.id = '...'
```

same for another column
```python
print(model_one.model_two.column_four)
```
it will the SQL to fetch that single column
```sql
SELECT model_two.column_four AS model_two_column_four 
FROM model_two 
WHERE model_two.id = '...'
```

Our use case is that for a web application, to check for authentication/ACL, we're fetching the minimal amount of data of a user, but if we need to access another attribute, it means that we're probably showing many things about that user, and we'd like to load all the deferred column in a single query.

In other words, after the first query I shared above, I'd like this to happen:
```python
print(model_one.model_two.column_one)  # no query, already loaded
print(model_one.model_two.column_three)  # queries all missing deferred columns
print(model_one.model_two.column_four)  # no query, already loaded from line above
```

Is there any way to do that?

Our current query looks like this
```python
model_one = (
    session.query(ModelOne)
    .filter(ModelOne.id == model_one_id)
    .options(
        joinedload(ModelOne.model_two)
    )
    .first()
)
```
and we're trying to optimize. The `model_one` is used in many places in the app (that's relatively big). Ideally, changes should happen in the query or just after, like
```python
model_one = (session.query(ModelOne)......)
session.something_magic(model_one)
print(model_one.model_two.column_one)
print(model_one.model_two.column_three)
print(model_one.model_two.column_four)
```
and not later. In other words, not this:
```python
print(model_one.model_two.column_one)
session.something_magic(model_one) # not here
print(model_one.model_two.column_three)
print(model_one.model_two.column_four)
```

Also, `model_one.model_two.column_one` is still valid, so I'm not looking to expire the whole object. I'd just want to load all deferred columns if any of them is accessed.

I hope my question is clear. Let me know if I can clarify. Thanks in advance!

**Comment by zzzeek** [comment]
there's a feature that sort of does this called "undefer group" but it is mostly tested only for columns that are deferred at the mapper level; if I set up this "group" for columns that are not normally deferred, ~~it is not working correctly~~ seems to work for 2.0 mappings, see next answer

you would need to issue a `session.refresh()` operation or a session.get() with populate_existing to instruct the object to load all the unloaded columns right now.   there's no option for load_only() to set some kind of "undefer everything on load" feature right now.

**Reply by zzzeek** [reply, marked as answer]
using 2.0 mappings this case seems to work , so I'm not sure what is not working for the legacy case, try this out

```py
from sqlalchemy import *
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import *
Base = declarative_base()

class ModelOne(Base):
    __tablename__ = "model_one"
    id = Column(String(32), primary_key=True)
    model_two_id = Column(
        String(32),
        ForeignKey("model_two.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    model_two = relationship("ModelTwo")

class ModelTwo(Base):
    __tablename__ = "model_two"
    id: Mapped[str] = mapped_column(primary_key=True)
    column_one: Mapped[str] = mapped_column(deferred_group="g1")
    column_two: Mapped[str] = mapped_column(deferred_group="g1")
    column_three: Mapped[str] = mapped_column(deferred_group="g1")
    column_four: Mapped[str] = mapped_column(deferred_group="g1")

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(ModelOne(id="m1", model_two=ModelTwo(id="m2", column_one='x', column_two='y', column_three='q', column_four='g')))
s.commit()


model_one = (
    s.query(ModelOne)
    .filter(ModelOne.id == "m1")
    .options(
        joinedload(ModelOne.model_two).options(
            load_only(
                ModelTwo.id,
                ModelTwo.column_one,
                ModelTwo.column_two,
            )
        )
    )
    .first()
)

print("-----------------")
print(model_one.model_two.column_three)

print("-----------------")
print(model_one.model_two.column_four)

```

output:

```
2023-11-30 10:00:47,325 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-11-30 10:00:47,327 INFO sqlalchemy.engine.Engine SELECT model_one.id AS model_one_id, model_one.model_two_id AS model_one_model_two_id, model_two_1.id AS model_two_1_id, model_two_1.column_one AS model_two_1_column_one, model_two_1.column_two AS model_two_1_column_two 
FROM model_one LEFT OUTER JOIN model_two AS model_two_1 ON model_two_1.id = model_one.model_two_id 
WHERE model_one.id = ?
 LIMIT ? OFFSET ?
2023-11-30 10:00:47,327 INFO sqlalchemy.engine.Engine [generated in 0.00008s] ('m1', 1, 0)
-----------------
2023-11-30 10:00:47,328 INFO sqlalchemy.engine.Engine SELECT model_two.column_one AS model_two_column_one, model_two.column_two AS model_two_column_two, model_two.column_three AS model_two_column_three, model_two.column_four AS model_two_column_four 
FROM model_two 
WHERE model_two.id = ?
2023-11-30 10:00:47,328 INFO sqlalchemy.engine.Engine [generated in 0.00007s] ('m2',)
q
-----------------
g

```

**Reply by zzzeek** [reply]
oh, OK.  above, using `deferred_group` means that those columns are now marked as deferred globally.  meaning if you didn't use load_only(), they would not load at all unless accessed.     so that's what we have at the moment, if columns are deferred at the mapper level, they can be part of a "group" that loads all at once.    load_only() still loads those individual columns but if you didnt' use load_only(), they'd stay deferred.

**Reply by antoineleclair** [reply]
Wow yes, I think that would solve it. In a perfect world, it would have been great to be able to define those groups at query time instead of directly in the model, so that other queries are not affected, but we can work around that :+1:

That's one good reason to update to 2.0 now.

Thanks!

**Reply by zzzeek** [reply]
you can get the same effect in 1.4 if you define all those columns using [deferred](https://docs.sqlalchemy.org/en/14/orm/loading_columns.html#deferred-column-loading)  .   it's the same configuration

**Reply by zzzeek** [reply]
like this

```py
class ModelTwo(Base):
    __tablename__ = "model_two"
    id: Mapped[str] = mapped_column(primary_key=True)
    column_one= deferred(Column(String(64)), group="g1")
    column_two= deferred(Column(String(64)), group="g1")
    column_three= deferred(Column(String(64)), group="g1")
    column_four= deferred(Column(String(64)), group="g1")

```

what we dont support right now is setting up that "group" with columns that are not mapper-level deferred

**Reply by antoineleclair** [reply]
Thank you for clarifying!
